### PR TITLE
Monster weight / volume granularity

### LINKED ISSUE #24909: Butchery system does not scale down properly to accommodate smaller monsters
The butchery system seems to neglect enemy sizes, or at least not scale properly.
1) Small enemies like rats, snakes, or baby spiders yield full-size corpses (30L, 50kg).
2) Overkill on enemies similarly yields excessive meat. (Blowing apart a baby spider yields a shower of 10 chunks of meat. A sewer snake yields 14 chunks of meat)

### LINKED ISSUE #24921: proposal for better granularity of meat yields
See #24909 

A creature's mass should be specified in its definition and on death a corpse item with the same mass should be generated. The yield of butchering a corpse should be derived from its mass.

This would allow granularity in meat yields that isn't constrained by size categories. The current system of deriving meat yields from animal size leads to some weird inconsistencies in yields and low granularity.

### LINKED ISSUE #25298: Size / Weight / Volume of monsters granularity
**Is your feature request related to a problem? Please describe.**  

This problem emerged along with introduction of non-standard creatures like dog breeds, and re-evaluation of meat yield per creature size. Those two changes resulted it some non-optimal meat yield counts for creatures that should not be that big/small, and in consequence showed that existing granularity of interdependent Size / Weight / Volume has reach it's limits.

**Describe the solution you'd like:**  
 - Add  "in-between" sizes and support for them,
- Re-balance whatever an wherever is needed,
- Reapply sizes to JSON-ized creatures.

| SIZE | REFERENCE | WEIGHT | CORPSE VOLUME |  MAX MEAT CHUNKS |
|------------|-------------|------------|-------------|--------------|
| **MS_TINY** |  Squirrel, snake | 1000_gram | 750_ml | 2 |
| MS_VERY_SMALL | Dog: dachshund |10000_gram | 7500_ml | 16 |
| **MS_SMALL** | Dog: labrador | 40750_gram | 30000_ml | 64 |
| MS_SMALL_MEDIUM | Dog: rottweiler | 60000_gram | 46250_ml | 96 |
| **MS_MEDIUM** | Human | 81500_gram | 62500_ml | 128 |
| MS_MEDIUM_LARGE | Bloated zombie | 100000_gram | 77500_ml | 160 |
| **MS_LARGE** | Cow | 120000_gram | 92500_ml | 192 |
| MS_EXTRA_LARGE | Zombie hulk, moose | 160000_gram | 483750_ml | 256 |
| **MS_HUGE** | Tank Drone | 200000_gram | 875000_ml | 320 |

**Disclaimer:** above values were not changed for existing sizes written in **bold**. Values for new "in-between" sizes are rough in-between values of existing sizes. While weight and dependent meat count sounds ok, the corpse volume sounds off even with existing values (human is 15 liter? cow is 23 liters?)

**Describe alternatives you've considered**  
I leave the alternatives to be discussed here.

**Additional context**  
Reference: #24909, #24921, #25244 & #25248.

### PR DESCRIPTION
**Pull request by nexusmrsep**
#### Summary
```SUMMARY: Infrastructure "Converts monster size to weight and volume."```

#### Purpose of change
Resolves #25298 Resolves #24921 Resolves #24909 
Many game features use monster size classes to determine some effects. While this is a decent method for some cases, in other, like butchery, makes proper balancing impossible because lack of granularity.

#### Describe the solution
This PR does following things:
- introduces `weight` and `volume` variables to monsters in `mtype`
- introduces automatic conversion of `volume` to `size` class via `volume_to_size` method
Conversion table below shows how middle values between existing hardcoded weights and volumes (also presented below) were used to determine delimiter (borders) between size classes.

| SIZE | REFERENCE |  CORPSE VOLUME | 
|------------|-------------|------------|
| MS_TINY |  Squirrel, snake | 750_ml |
| **BORDERLINE** |------------ | **7500_ml** |
| MS_SMALL | Dog | 30000_ml |
| **BORDERLINE** | ------------- | **46250_ml** |
| MS_MEDIUM | Human | 62500_ml |
| **BORDERLINE** | --------------- | **77500_ml** |
| MS_LARGE | Cow |  92500_ml |
| **BORDERLINE** | -------------- | **483750_ml** |
| MS_HUGE | Tank Drone | 875000_ml |

- removes decoding `size`  entries from json monster files and add decoding `volume` and `weight` from them
- in JSON files for monsters replaces `"size"` with `"volume"` and`"weight"`, that were set at previously hardcoded weights and volumes (presented above)
This of course asks for a general refactor of those values according to some more sane values in many cases (example: most dogs weight over 40 kg)
- method `get_meat_chunks()` that is a primary function to determine main butcher yields was redesigned to use monster's `weight` instead of size class. It has also been given dependency on 'meat' type to dynamically resolve number of maximum yielded items as some creatures yield not "meat" per se, but thinks like "tainted meat", "veggies", etc. that might have different weight. In general this new calculation follows @mlangsdorf idea for a function from #25298 , but I've lowered its base value and flattened its regression, to keep it around 33% of body mass. It's final (?) form is:
 **weight * ( 0.4 - 2 * ln( weight ) )** divided by mass of a proper 'meat chunk'
Typical results in %:

| WEIGHT in kg | YIELD in % body mass | MEAT count |
| ----- | -------- | ----------- |
| 1 | 40%  | 2 |
| 5 | 36.7% | 9 |
| 10 | 35.3% | 18 |
| 20 | 34% | 34 |
| 50 | 32.1% | 80 |
| 100 | 30.8% | 154 |
| 200 | 29.4% | 294 |
| 300 | 28.5% | 429 |

This is of course dynamic, so different weights of creatures will result in granular effects.

#### Describe alternatives you've considered
Alternative was granulating size classes, even after introducing this PR, but a need for that would have to emerge, and there are too many places where they are uses, that would need rebalancing.

#### Additional context
Reference: #24909, #24921, #25244 & #25248, #24878

#### WARNING

- untested - TEST BEFORE MERGING
- will affect unsupported mods to fix their monster json files to use weight and volume instead of size

### CODE REVIEW

**Comment by SunshineDistillery** [comment]
Does it check the monster's material? Non-meat monsters should have different density/weight.

**Comment by nexusmrsep** [comment]
@SunshineDistillery this is an infrastructure conversion for the starters, any re-balancing and improvement will need to be done later.

**Comment by brianlewisbeck** [comment]
mtype::get_meat_chunks() forced corpses into 5 possible sizes.  All corpses of the smallest size (tiny) were 1 kg; all corpses of the next smallest size (small) were 40 kg.  There is a huge gap between 1 and 40kg.

**Comment by nexusmrsep** [comment]
@brianlewisbeck I have no idea what are you trying to tell here. This PR is all about changing that so... whats your point?

**Comment by brianlewisbeck** [comment]
Sorry - guess I was thinking out loud.  Does the issue relate to any of the other sizes?  If not, would it be possible to keep the existing framework but add subcategories to small?  Seems like that would be a lot easier.

**Comment by nexusmrsep** [comment]
> Does the issue relate to any of the other sizes? If not, would it be possible to keep the existing framework but add subcategories to small? Seems like that would be a lot easier.

To put it simple: when coding a new monster in JSON you will be able to define its weight and volume instead of forcing it into any constraints. For example you could say your new dog breed now weights 7 kg and is 15 liters of volume. Specific in game calculations would then use that value - for example to calculate meat chunks count, or if it just needs a category of size it would use a size class like before, but that class would be calculated from the given volume - for ex. 15 liter dog falls into MS_SMALL category.

Adding subcategories is not an option now - would require to rebalance "half" of the game.

**Comment by mlangsdorf** [comment]
is there a plan to replace the existing calls to get_size() where appropriate? Replacing the call in vehicle_move.cpp:part_collision() to critter->get_size() and the resulting switch/case set to get critter mass with get_weight(), for instance?

Or do you just plan to do this infrastructure and let the rest of us sort it out, which is fine, just want to know the plan.

**Comment by nexusmrsep** [comment]
If there is an obvious function that uses get   _size while it should use true weight or volume then yes. However the only one that I considered obvious was butchery yields up until now. If it would be intuitive for me to incorporate the one you mentioned then I'll try, but if not I'll pass. I will not change or balance creatures’ values either, just the infrastructure for it.

Second thing is that my free coding time got extremely scarce lately and I've got two other WIP not yet published PRs on the bench that slow me down in finishing this one. I just hope I'm closing in on finishing them.

**Comment by mlangsdorf** [comment]
I listed every use of MS_SIZE I could find in the code in my first response in #25298. But if you're too busy, I totally understand and can clean up the vehicles code on my own after you've got this merged. Like I wrote, I don't hugely care what the answer is, I just want to know the plan. If you're only going to make the infrastructure and butchery changes, that's fine and I won't ask you to change the vehicles code if I review this PR.

**Comment by nexusmrsep** [comment]
I'm removing [WIP] from this PR but please TEST BEFORE MERGING.

File: data/json/monsters.json [hunk 1]
```diff
@@ -7,7 +7,8 @@
     "default_faction": "",
     "species": [ "ZOMBIE" ],
     "diff": 100,
-    "size": "MEDIUM",
+    "volume": "62500 ml",
```
Comment (ZhilkinSerg) [inline data/json/monsters.json]: Why not `"volume": 62500` instead of `"volume": "62500 ml"`?

File: data/json/monsters.json [hunk 2]
```diff
@@ -7,7 +7,8 @@
...
+    "weight": "81500",
```
Comment (ZhilkinSerg) [inline data/json/monsters.json]: Why not `"weight": 81500` instead of `"weight": "81500"`?

File: data/json/monsters.json (hunk 1 above)
Comment (nexusmrsep) [reply data/json/monsters.json]: After Discord discussion this one stays, as this format is expected by assign() function.

File: data/json/monsters.json (hunk 2 above)
Comment (nexusmrsep) [reply data/json/monsters.json]: And in this I will regex quote marks into oblivion ASAP.

File: src/game_constants.h [hunk 3]
```diff
@@ -138,4 +138,7 @@ constexpr double accuracy_grazing  = 1.0;
 /** New England base spring temperature = 65F/18C rounded to int */
 #define SPRING_TEMPERATURE 65
 
+/** Character's base weight in gram */
```
Comment (BevapDin) [inline src/game_constants.h]: Why not in `units::mass`?

File: src/monstergenerator.cpp [hunk 4]
```diff
@@ -547,8 +548,10 @@ void mtype::load( JsonObject &jo, const std::string &src )
     }
 
     assign( jo, "color", color );
-    const typed_flag_reader<decltype( Creature::size_map )> size_reader{ Creature::size_map, "invalid creature size" };
-    optional( jo, was_loaded, "size", size, size_reader, MS_MEDIUM );
+    //const typed_flag_reader<decltype( Creature::size_map )> size_reader{ Creature::size_map, "invalid creature size" };
```
Comment (BevapDin) [inline src/monstergenerator.cpp]: Please do not check in commented out code.

File: src/mtype.cpp [hunk 5]
```diff
@@ -52,6 +57,23 @@ std::string mtype::nname( unsigned int quantity ) const
     return ngettext( name.c_str(), name_plural.c_str(), quantity );
 }
 
+m_size mtype::volume_to_size( const units::volume vol ) const
```
Comment (BevapDin) [inline src/mtype.cpp]: Why is this a non-static functions? It does not access any of the members of `mtype`, it does not needthe implicit `this` pointer. Make it a static function.

File: src/mtype.cpp [hunk 6]
```diff
@@ -52,6 +57,23 @@ std::string mtype::nname( unsigned int quantity ) const
...
+{
+    if( vol > 0_ml && vol <= 7500_ml ) {
+        return MS_TINY;
+    } else if( vol > 7500_ml && vol <= 46250_ml ) {
```
Comment (BevapDin) [inline src/mtype.cpp]: You don't need the first check, write it like this instead:
```C++
if( vol <= 7500_ml ) {
    return ...;
} else if( vol <= 46250_ml ) {
    return ...;
} else if( vol <= ...
```
The second if clause will only be reached if `vol > 7500_ml`, which does not need to be cecked again.

File: src/mtype.cpp [hunk 7]
```diff
@@ -200,19 +222,9 @@ itype_id mtype::get_meat_itype() const
...
-            return 128;
-        case MS_LARGE:
-            return 192;
-        case MS_HUGE:
-            return 320;
-    }
-    return 0;
+    float ch = units::to_gram( weight ) * ( 0.40f - 2 * log( units::to_gram( weight ) ) );
+    const itype *chunk = item::find_type( get_meat_itype() );
+    return static_cast<int>( ch / units::to_gram( chunk->weight ) );
```
Comment (BevapDin) [inline src/mtype.cpp]: `units::` is not needed when calling `to_gram` with a `units::mass` argument, the C++ compiler will find it anyway thanks to ADL.

File: src/creature.h [hunk 8]
```diff
@@ -55,6 +55,8 @@ template<typename V, typename U>
 class quantity;
 class mass_in_gram_tag;
 using mass = quantity<int, mass_in_gram_tag>;
+class volume_in_milliliter_tag;
+using volume = quantity<int, volume_in_milliliter_tag>;
```
Comment (BevapDin) [inline src/creature.h]: And why did you add those two lines?

File: src/game_constants.h (hunk 3 above)
Comment (nexusmrsep) [reply src/game_constants.h]: no reason other then not including units.h, but i guess there is no difference, and it's more clear using the suggested way

File: src/monstergenerator.cpp (hunk 4 above)
Comment (nexusmrsep) [reply src/monstergenerator.cpp]: good call, didn't even noticed that I left it behind

File: src/mtype.cpp (hunk 5 above)
Comment (nexusmrsep) [reply src/mtype.cpp]: ok, this requires moving it to `monstergenerator.cpp`, I thought it might be more universal in the future by placing it in `mtype`, but lets leave the future to govern itself.

File: src/mtype.cpp (hunk 6 above)
Comment (nexusmrsep) [reply src/mtype.cpp]: I just hope there will be no situation where `vol == 0` then...

File: src/creature.h (hunk 8 above)
Comment (nexusmrsep) [reply src/creature.h]: Notice line 56-57 where similar lines for mass were already added?
It prevents this compile error:
image

File: src/monstergenerator.cpp [hunk 9]
```diff
@@ -102,6 +102,20 @@ static int calc_bash_skill( const mtype &t )
...
+{
+    if( vol <= 7500_ml ) {
+        return MS_TINY;
+    } else if( vol <= 46250_ml ) {
+        return MS_SMALL;
+    } else if( vol <= 77500_ml ) {
+        return MS_MEDIUM;
+    } else if( vol <= 483750_ml ) {
+        return MS_LARGE;
+    }
```
Comment (ZhilkinSerg) [inline src/monstergenerator.cpp]:
```suggestion
    }
```

**Comment by ZhilkinSerg** [comment]
Great. Will test it again tonight.

**Comment by GoLoT** [comment]
I'm not sure if this is the place or the way to report it but this commit has broken butchering completely. I can't debug because the latest experimental refuses to compile under VS but I have taken a look at the commit and I think it has to do with how the mass percent calculation is done:

`float ch = to_gram( weight ) * ( 0.40f - 2 * log( to_gram( weight ) ) );`
This line seems to be using weight in grams instead of kilograms which causes the calculations to be way off. The percentage calculation also seems wrong as 2*log(w) should be substracted from 40 and not 0.4, causing ch to always be a negative number and making every butcher attempt fail.

A quick fix that I can't test at the moment could be:
`float ch = to_kilogram( weight ) * ( 0.40f - 0.02 * log( to_kilogram( weight ) ) );`

And then divide by the chunk weight in kilograms:
`return static_cast<int>( ch / to_kilogram( chunk->weight ) );`

File: src/mtype.cpp [hunk 10]
```diff
@@ -200,19 +205,9 @@ itype_id mtype::get_meat_itype() const
...
-            return 64;
-        case MS_MEDIUM:
-            return 128;
-        case MS_LARGE:
-            return 192;
-        case MS_HUGE:
-            return 320;
-    }
-    return 0;
+    float ch = to_gram( weight ) * ( 0.40f - 2 * log( to_gram( weight ) ) );
```
Comment (ZhilkinSerg) [inline src/mtype.cpp]: Should be:

```CPP
float ch = to_gram( weight ) * ( 0.40f - 0.02f * log10( to_gram( weight ) ) );
```

Comment (nexusmrsep) [reply src/mtype.cpp]: Todo ASAP today.

**Comment by jbytheway** [comment]
See also #26450.
